server_client/src/renderer/LeftPanel.tsx

```diff
@@ -229,56 +230,71 @@ const Buttons = (props: {
   );
 };
 
-export const LeftPanel = (props: { pages: Page[] }) => {
-  const { pages } = props;
+export const LeftPanel = (props: {
+  pages: Page[];
+  setPages: React.Dispatch<React.SetStateAction<Page[] | null>>;
+}) => {
+  const { pages, setPages } = props;
   const [stretchButtons, setStretchButtons] = useState(false);
   const [colWidth, setColWidth] = useState(150);
   const [rowHeight, setRowHeight] = useState(150);
-  const [pageHeight, setPageHeight] = useState(0);
-  const [minPageHeight, setMinPageHeight] = useState(0);
-  const [minPageWidth, setMinPageWidth] = useState(0);
-  const [pageWidth, setPageWidth] = useState(0);
+  const [minPageHeight, setMinPageHeight] = useState(1);
+  const [minPageWidth, setMinPageWidth] = useState(1);
   const [pageId, setPageId] = useState(pages[0].id);
   const [page, setPage] = useState(pages[0]);
 
+  useEffect(() => {
+    if (!pages || pages === null || pages === undefined) {
+      return;
+    }
+    const savedPage = pages.find((p) => p.id === page.id);
+    if (savedPage && !_.isEqual(savedPage, page)) {
+      const newPages = _.cloneDeep(pages);
+      const index = newPages.findIndex((p) => p.id === page.id);
+      if (index === -1) {
+        return;
+      }
+      newPages[index] = page;
+
+      setPages(newPages);
+    }
+  }, [page, pages, setPages]);
+
   useEffect(() => {
     const determineMinWidthHeight = (p: Page) => {
       let highestX = 0;
       let highestY = 0;
 
-      for (let y = 0; y < p.height - 1; y++) {
-        for (let x = 0; x < p.width - 1; x++) {
-          const button = p.buttons.find((b) => b.x === x && b.y === y);
-          if (button) {
-            if (y > highestY) {
-              highestY = y;
-            }
-            if (x > highestX) {
-              highestX = x;
-            }
-          }
+      p.buttons.forEach((button: Button) => {
+        if (button.y > highestY) {
```

Comment:
Unrelated to pr, but optimises height/width bounds to only check buttons, not the whole grid.